### PR TITLE
Decouple the graph and render logic from the fs watcher

### LINKED ISSUE #1891: watching newly created files
Greetings!
When I start node-sass in recursive watching mode it works just fine with the exception that newly created files are not being watched.
For example, if I wach a directory and I add new _styles.scss_ file, node-sass ignores this file, providing the process has already been started before. 
Is it possible to change this behavior? Or do I just miss something?

### PR DESCRIPTION
**Pull request by xzyfer**
This logic is all tightly coupled in the bin. This logic has been
notoriously impossible to test due to weird fs timing issues. By
extracting the actual logic we're now able to test it in isolation.

With this in place replacing Gaze (#636) becomes a viable option.
This PR not only massively increases our test coverage for the
watcher but also address a bunch of known edge cases i.e. orphaned
imports when a files is deleted.

Closes #1896
Fixes #1891

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -75,12 +75,14 @@
   "devDependencies": {
     "coveralls": "^2.11.8",
     "eslint": "^3.4.0",
+    "fs-extra": "^0.30.0",
```
Comment (xzyfer) [inline package.json]: This is a super older version but it's the last version that supports Node 0.10.

**Comment by xzyfer** [comment]
>Abstract all the gaze.xxx calls in bin/node-sass to functions in the watcher

Can you elaborate on how this differs from what I've done here?

>Extract the watcher to a new NPM module 

This is related to #1056. I'm open to it.

File: bin/node-sass [hunk 2]
```diff
@@ -229,63 +229,41 @@ function getOptions(args, options) {
...
-    };
-
-    if (options.directory) {
-      graph = grapher.parseDir(options.directory, graphOptions);
-    } else {
-      graph = grapher.parseFile(options.src, graphOptions);
-    }
+  var handler = function(files) {
+    files.added.forEach(function(file) {
+      var watch = gaze.watched();
```
Comment (nschonni) [inline bin/node-sass]: Abstract was probably the wrong word, but here is a crude example example
```js
    var watch = watcher.watched()
      if (watch.indexOf(file) === -1) {
        watcher.add(file);
      }
    });
```
So the `watcher` module would export certain functions required in the bin

**Comment by nschonni** [comment]
The other part of splitting it would be a conditional `require`
```
var watcher;
if (option.watcher !== null)
  watcher = require(option.watcher)
 // error handling
} else {
  watcher = require('node-sass-watcher');
}

File: bin/node-sass (hunk 2 above)
Comment (xzyfer) [inline bin/node-sass]: Sorry I'm still not following. 

If I'm understanding correctly wouldn't that then couple us to the fs watcher implementation's api? 
What happens if the next fs watcher doesn't expose a method like `watch()` for exposing all currently watched files?

The `watch.indexOf(file) === -1` is an optimisation to prevent calling `gaze.add` superfluously. Under the covers `gaze.add` also does deduping but there's not guarantee the next fs watcher would.

File: bin/node-sass [hunk 3]
```diff
@@ -229,63 +229,41 @@ function getOptions(args, options) {
...
+      if (watch.indexOf(file) === -1) {
+        gaze.add(file);
+      }
+    });
 
-    return graph;
+    files.changed.forEach(function(file) {
+      if (path.basename(file)[0] !== '_') {
```
Comment (xzyfer) [inline bin/node-sass]: We could move this logic into `lib/watcher` but if we were to extract it into it's own package I can see uses cases for people wanting to poke changed partials i.e. `url` rewriting, or some kind webpack magic.

File: bin/node-sass [hunk 4]
```diff
@@ -229,63 +229,41 @@ function getOptions(args, options) {
...
+        renderFile(file, options, emitter);
+      }
+    });
+
+    files.removed.forEach(function(file) {
```
Comment (xzyfer) [inline bin/node-sass]: This could be

```
files.removed.forEach(gaze.remove)
```

but it's not likely the next watcher would have the same API

**Comment by nschonni** [comment]
I think this is probably fine to land as-is as it gets us testing.
Looking at the `bin` again, if we do a second pass, I think all of the gaze/watcher may just be pulled out by changing https://github.com/xzyfer/node-sass/blob/e81d6b9f225be4ea124a244d6e287a3711044e9f/bin/node-sass#L312
```js
   if (options.watch) {
     watcher.handle(options, emitter);
   } else if (options.directory) {
     renderDir(options, emitter);
   } else {
     render(options, emitter);
}
```
Although I may be missing an obvious problem with passing it that way

**Comment by xzyfer** [comment]
Yeah there will be more passes. I also want to extract the console emiiter stuff because it's the cause of a lot of issues and feature request.
